flink-state-backends/flink-statebackend-heap-spillable/src/test/java/org/apache/flink/runtime/state/heap/CopyOnWriteSkipListStateMapTest.java

```diff
@@ -115,577 +144,783 @@ public void testInitStateMap() {
-						stateMap.remove(key, namespace);
-					} else {
+					removedCnt++;
+					// remove from state map
+					if (getOld) {
 						assertEquals(state, stateMap.removeAndGetOld(key, namespace));
+					} else {
+						stateMap.remove(key, namespace);
 					}
-					assertEquals(totalSize, stateMap.size());
-					assertEquals(totalSize, stateMap.totalSize());
+					// remove from reference to keep in accordance
+					kvIterator.remove();
 				}
 			}
 		}
+		return removedCnt;
+	}
 
-		for (Map.Entry<Long, Set<Integer>> entry : removedStates.entrySet()) {
-			long namespace = entry.getKey();
-			Map<Integer, String> keyMap = referenceStates.get(namespace);
-			if (keyMap != null) {
-				entry.getValue().forEach(keyMap::remove);
-				if (keyMap.isEmpty()) {
-					referenceStates.remove(namespace);
+	private int removeAbsentState(
+			CopyOnWriteSkipListStateMap<Integer, Long, String> stateMap,
+			Map<Long, Map<Integer, String>> referenceStates) {
+		return applyFunctionAfterRemove(
+			stateMap,
+			referenceStates,
+			(removedCnt, removedStates) -> {
+				// remove the same keys again, which would be absent already
+				for (Map.Entry<Long, Set<Integer>> entry : removedStates.entrySet()) {
+					long namespace = entry.getKey();
+					for (int key : entry.getValue()) {
+						assertNull(stateMap.removeAndGetOld(key, namespace));
+					}
 				}
+				return removedCnt;
 			}
-			for (int key : entry.getValue()) {
-				assertNull(stateMap.get(key, namespace));
-				assertFalse(stateMap.containsKey(key, namespace));
-			}
-		}
-
-		assertEquals(totalSize * 2, spaceAllocator.getTotalSpaceNumber());
-		verifyState(referenceStates, stateMap);
+		);
+	}
 
-		// update some states
+	/**
+	 * Apply the given function after removing some states.
+	 *
+	 * @param stateMap the state map to test against.
+	 * @param referenceStates the reference of states for correctness verfication.
+	 * @param function a {@link BiFunction} which takes [removedCnt, removedStates] as input parameters.
+	 * @param <R> The type of the result returned by the function.
+	 * @return The result of applying the given function.
+	 */
+	private <R> R applyFunctionAfterRemove(
+		CopyOnWriteSkipListStateMap<Integer, Long, String> stateMap,
+		@Nonnull Map<Long, Map<Integer, String>> referenceStates,
+		BiFunction<Integer, Map<Long, Set<Integer>>, R> function) {
+		int removedCnt = 0;
+		Map<Long, Set<Integer>> removedStates = new HashMap<>();
+		// remove some state
 		for (Map.Entry<Long, Map<Integer, String>> namespaceEntry : referenceStates.entrySet()) {
 			long namespace = namespaceEntry.getKey();
-			for (Map.Entry<Integer, String> keyEntry : namespaceEntry.getValue().entrySet()) {
+			Map<Integer, String> kvMap = namespaceEntry.getValue();
+			Iterator<Map.Entry<Integer, String>> kvIterator = kvMap.entrySet().iterator();
+			while (kvIterator.hasNext()) {
+				Map.Entry<Integer, String> keyEntry = kvIterator.next();
 				if (random.nextBoolean()) {
```

Comment:
If something fails for whatever reason, this `random` behaviour will make it super hard to reproduce it.